src/workload/mod.rs

```diff
@@ -727,7 +778,123 @@ impl Keyspace {
+        let mut raw_seed = [0_u8; 64];
+        rng.fill_bytes(&mut raw_seed);
+
+        // we use a predictable seed to generate the keys in the store
+        let mut rng = Xoshiro512PlusPlus::from_seed(key_seed);
+        let mut keys = HashSet::with_capacity(nkeys);
+        while keys.len() < nkeys {
+            let key = (&mut rng)
+                .sample_iter(&Alphanumeric)
+                .take(klen)
+                .collect::<Vec<u8>>();
+            let _ = keys.insert(key);
+        }
+        let keys = keys.drain().map(|k| k.into()).collect();
+        let key_dist = match store.key_distribution() {
+            config::Distribution::Uniform => Distribution::Uniform(Uniform::new(0, nkeys)),
+            config::Distribution::Zipf => {
+                Distribution::Zipf(ZipfDistribution::new(nkeys, 1.0).unwrap())
+            }
+        };
+
+        let mut commands = Vec::new();
+        let mut command_weights = Vec::new();
+
+        for command in store.commands() {
+            commands.push(*command);
+            command_weights.push(command.weight());
+
+            // validate that the store is adaquately specified for the given
+            // verb
+
+            // commands that set generated values need a `vlen`
+            if store.vlen().is_none()
+                && store.vkind() == ValueKind::Bytes
+                && matches!(command.verb(), StoreVerb::Set)
+            {
+                eprintln!(
+                    "verb: {:?} requires that the keyspace has a `vlen` set when `vkind` is `bytes`",
+                    command.verb()
+                );
+                std::process::exit(2);
+            }
+        }
+
+        let command_dist = WeightedAliasIndex::new(command_weights).unwrap();
+        Self {
+            keys,
+            key_dist,
+            commands,
+            command_dist,
+            vlen: store.vlen().unwrap_or(0),
+            vkind: store.vkind(),
+            value_random_bytes,
+        }
+    }
+
+    pub fn sample(&self, rng: &mut dyn RngCore) -> Arc<[u8]> {
+        let index = self.key_dist.sample(rng);
+        self.keys[index].clone()
+    }
+
+    pub fn sample_string(&self, rng: &mut dyn RngCore) -> Arc<String> {
+        let keys = self.sample(rng);
+        Arc::new(String::from_utf8_lossy(&keys).into_owned())
+    }
+
+    pub fn gen_value(&self, rng: &mut dyn RngCore) -> Vec<u8> {
+        match self.vkind {
+            ValueKind::I64 => format!("{}", rng.gen::<i64>()).into_bytes(),
+            ValueKind::Bytes => {
+                let mut buf = vec![0_u8; self.vlen];
+                rng.fill(&mut buf[0..self.value_random_bytes]);
+                buf
+            }
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub enum ClientWorkItemKind<T> {
+    Reconnect,
+    Request { request: T, sequence: u64 },
+}
+
+pub async fn reconnect<TRequestKind>(
+    work_sender: Sender<ClientWorkItemKind<TRequestKind>>,
+    config: Config,
+) -> Result<()> {
```

Comment:
This is the impetus for DRY'ing up the `WorkItem` types. It seemed unnecessary to copy and paste this entire sequence of code to support a Store client when the end result and means is the exact same.

The only difference between the two is the inner request type, which can be specified at compile-time (as seen in changes throughout the codebase).